SetProtocol/set-protocol-v2
# feat(publishing): Auto publish repository to npm in CI [SIM-123]

+ Adds [semantic-release][2] package and logic to publish the repo in CI on merge to master. 
+ Adds npm commands to publish different builds from the command line
  + `publish:local:latest` (for regular releases)
  + `publish:local:hardhat` (for the release we consume at set-v2-strategies)
+ Updates lowest supported Node version to 14 (necessary to install `semantic-release`)
+ Update typechain/hardhat and related deps to `V6` (necessary to generate `.d.ts` files in the dist/typechain folder) 
+ Adds bot to CI to enforce auto-release PR title semantics (see below)
+ Add `GITHUB_TOKEN` and `NPM_TOKEN` personal access token environment variables to CircleCI env

We currently publish two builds for internal  use:
+ `latest` (consumed at set.js and the deployments repos) which has conventional hardhat artifacts and deposits TS definitions into a dist/types folder
+ `hardhat` (consumed at set-v2-strategies) which hardcodes gas values into the artifact abis and deposits TS types alongside the transpiled files. (Both of these are necessary to make the Perp fixtures imported at set-v2-strategies to work correctly)

To trigger an automatic release on merge to master, the following naming convention must be used in a PR title (or in the squash & merge commit description). 

| Type | Increment | Pattern | Example |
| ---- | ---- | ---- | ---- |
| patch | 1.0.1 -> 1.0.2 | *fix(topic): comment* | *fix(PerpV2Viewer): return uint256 instead of int256* |
| minor | 1.1.1 -> 1.2.0 | *feat(feature_name): comment* | *feat(BasisTrading): Add PerpV2BasisTradingModule* |


**Bot screenshot**

![Screen Shot 2022-03-15 at 8 15 44 AM](https://user-images.githubusercontent.com/7332026/158410607-63cdc6db-d19b-49ed-be86-94bf2bf5476e.png)


**TODO**
- [x] Check that hardhat prerelease installs and runs correctly at set-v2-strategies
  + [set-v2-strategies PR 20][1]
- [x] Make sure package.json version field is updated
- [x] Delete test releases from Github
- [x] Add [internal gitbook documentation][3] about this and versioning policy 

[1]: https://github.com/SetProtocol/set-v2-strategies/pull/20
[2]: https://semantic-release.gitbook.io/semantic-release/v/beta/
[3]: https://app.gitbook.com/o/-MGdl9Y5UCSpZPXC3ad7/s/-MGdlDDRsIRuOJOl7btN/function-tools-and-guides/engineering/development-guides/smart-contract/how-tos/publishing
